compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2876,6 +2876,12 @@ public void visit(BLangSimpleVarRef varRefExpr, AnalyzerData data) {
                 varRefExpr.symbol = varSym;
                 actualType = varSym.type;
                 markAndRegisterClosureVariable(symbol, varRefExpr.pos, data.env, data);
+
+                if (varRefExpr.isLValue && (varSym.tag & SymTag.FUNCTION) == SymTag.FUNCTION) {
+                    actualType = symTable.semanticError;
+                    dlog.error(varRefExpr.pos, DiagnosticErrorCode.INVALID_ASSIGNMENT_DECLARATION_FINAL,
+                            Names.FUNCTION);
+                }
```

Comment:
I think it is better to do this check in https://github.com/ballerina-platform/ballerina-lang/blob/acd3357a4fe56a4f01fb85d942f499795cadf172/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java#L2256,
because we don't want to check for all the simple variable references.

Reply:
Here we first check for `varRefExpr.isLValue` and then we do the check, So we are not checking for all the simple variable references. IMO, both are same.

Reply:
Updated 648b6f5c2286d7e8281a917dfa543eb2e24448e1